Add blog post section template

tauthomas01 opened the pull request:
Reference: #8
[Demo store](https://oecry1oqn5m1gk8q-662601789.shopifypreview.com)

### WHAT are you trying to accomplish with this PR?
I am adding "blog-post" sections template from `Debut` theme as a starting point in order to see the newest change when I adapt to the `starter theme` approach.

### WHY are you choosing this approach?
In order to see the changes when I adapt the "starter theme" way, we need a reference to see what has been changed.

### HOW is this accomplished?
Copied the `featured-blog.liquid` from Debut theme

t-kelly commented on src/sections/blog-post.liquid:
```diff
@@ -0,0 +1,184 @@
+<div class="page-width">
```
^^^ Comment block describing:
- This is a required section for the Shopify Theme Store
- What this section does
- What are the required settings
- What are the optional settings

tauthomas01 replied:
I'm still working on the comment block so the text might change.

t-kelly commented on src/sections/blog-post.liquid:
```diff
@@ -0,0 +1,184 @@
+<div class="page-width">
```
Maybe we could use the `<section>` tag as a wrapper? No need for the class.

NathanPJF replied:
`<section>` for sure wrapping most (likely all) of these homepage sections. 

I had a chat with Scott about this PR as there are some general HTML structure questions I had.  I'll put them as a comment on this thread.

tauthomas01 replied:
I use `<section>` now

t-kelly commented on src/sections/blog-post.liquid:
```diff
@@ -0,0 +1,184 @@
+<div class="page-width">
+  {% if section.settings.title != blank %}
+    <header class="section-header text-center">
```
Might not need this wrapper `<header>` container

NathanPJF replied:
The `header` would need an associated "section-ing element", like `<body>`, `<section>`, `<article>`.

There's a suggestion to add one above, but I still agree with your comment.  This would be fine with the `<h2>` as there isn't much going on in this `<header>`.

useful source: https://internetingishard.com/html-and-css/semantic-html/#headers

NathanPJF replied:
@tauthomas01 something to be conscious of: Starter theme should be shipping with classes that are unused.  In this case `section-header`.  The `text-center` class is used in the base CSS, but I feel that should be removed as well.  No need to force a centered heading on folks as a starting point.

tauthomas01 replied:
Yes, I removed `<header>` since there are not much information to show.

t-kelly commented on src/sections/blog-post.liquid:
```diff
@@ -0,0 +1,184 @@
+  {% endif %}
+
+  {%- assign blog = blogs[section.settings.blog] -%}
+
```
Maybe a comment here about how the following Liquid is for when we actually have blog items?

tauthomas01 replied:
Added.

t-kelly commented on src/sections/blog-post.liquid:
```diff
@@ -0,0 +1,184 @@
+          </div>
+        </div>
+      {% endfor %}
+    </div>
```
And a comment here about how this Liquid below is for when no blog posts exist (place holder)

t-kelly commented on src/sections/blog-post.liquid:
```diff
@@ -0,0 +1,184 @@
+
+  {% unless blog.articles_count == 0 %}
+    <div class="grid grid--uniform grid--blog">
+      {% for article in blog.articles limit: section.settings.post_limit %}
```
Actually this can be simplified. We can remove the above `{% unless %}` statement and use this `{% for ... %}` statement with an [`{% else %}` tag](https://help.shopify.com/themes/liquid/tags/iteration-tags#else)

tauthomas01 replied:
I tried it but I realized we need a parent container to wrap the items when we show as a grid so unfortunately I can't do it :( Unless there's another solution I'm not aware of.

t-kelly commented on src/sections/blog-post.liquid:
```diff
@@ -0,0 +1,184 @@
+                </a>
+              </li>
+
+              {% if blog.comments_enabled? and article.comments_count > 0 %}
```
This if condition and its contents are probably not needed

NathanPJF replied:
@t-kelly You mean that Starter theme doesn't need to have this out of the box?  I'm a fan of having it 😃

tauthomas01 replied:
I removed it for now but I can revert back anytime.

t-kelly commented on src/sections/blog-post.liquid:
```diff
@@ -0,0 +1,184 @@
+    </div>
+  {% endunless %}
+
+  {% if section.settings.show_view_all %}
```
This if statement and its contents are probably not needed

tauthomas01 replied:
Removed.

NathanPJF commented on src/sections/blog-post.liquid:
```diff
@@ -0,0 +1,184 @@
+  {%- assign blog = blogs[section.settings.blog] -%}
+
+  {% unless blog.articles_count == 0 %}
+    <div class="grid grid--uniform grid--blog">
```
`grid--blog` can be removed.

tauthomas01 replied:
Yes, I removed it and created a custom stylesheet for blog post only. I'm applying CSS grids

NathanPJF commented on src/sections/blog-post.liquid:
```diff
@@ -0,0 +1,184 @@
+              </a>
+            </div>
+            <noscript>
+              <a href="{{ article.url }}" class="article__grid-image-wrapper">
```
Again, I would remove this class `article__grid-image-wrapper` and all other of these article specific BEM classes since they are pretty specific and we aren't shipping the matching CSS. 
Is this how you see it @t-kelly ?

tauthomas01 replied:
I removed.

NathanPJF commented on src/sections/blog-post.liquid:
```diff
@@ -0,0 +1,184 @@
+          {% endif %}
+
+          <div class="article__grid-meta{% if article.image %} article__grid-meta--has-image{% endif %}">
+            <h2 class="h3 article__title">
```
Should be an `h3` since the section is using `h2`

NathanPJF commented on src/sections/blog-post.liquid:
```diff
@@ -0,0 +1,184 @@
+
+            <ul class="list--inline article__meta-buttons">
+              <li>
+                <a href="{{ article.url }}" class="btn btn--secondary btn--small">
```
In an onboarding context, there will be no `article` object to reference here.

tauthomas01 replied:
What does "onboarding context" mean?

NathanPJF commented on src/sections/blog-post.liquid:
```diff
@@ -0,0 +1,184 @@
+
+              {% if blog.comments_enabled? and article.comments_count > 0 %}
+                <li>
+                  <a href="{{ article.url }}#comments" class="btn btn--small btn--link">
```
Unused CSS `btn btn--small btn--link`

tauthomas01 replied:
I removed

NathanPJF commented on src/sections/blog-post.liquid:
```diff
@@ -0,0 +1,184 @@
+            </h2>
+
+            {% if section.settings.blog_show_author %}
+              <span class="article__author">{{ 'blogs.article.by_author' | t: author: article.author }}</span>
```
For any of these translation filter `| t` you'll need matching keys in the locales file.  For example, the `blogs` object doesn't have a `by_author`

https://github.com/Shopify/starter-theme/blob/master/src/locales/en.default.json#L48-L52

tauthomas01 replied:
I added the key in every locale file, but I couldn't find for `da.json` and `nl.json`

t-kelly commented on src/sections/blog-post.liquid:
```diff
@@ -0,0 +1,155 @@
+  This is a required section for the Shopify Theme Store.
+  It is loaded when you add "Blog Posts" section on the Theme Editor.
+
+  We are showing a preview of what the blog post looks like in the preview.
```
Not sure what you mean by this.

tauthomas01 replied:
I removed it. After reading it, it doesn't make sense 😢

t-kelly commented on src/sections/blog-post.liquid:
```diff
@@ -0,0 +1,155 @@
+
+  We are showing a preview of what the blog post looks like in the preview.
+
+  Required settings
```
Maybe put `Theme Store required settings`?

tauthomas01 replied:
It's good.

t-kelly commented on src/sections/blog-post.liquid:
```diff
@@ -0,0 +1,155 @@
+
+            {% if article.tags.size > 0 %}
+              {% for tag in article.tags %}
+                <a href="{{ blog.url }}/tagged/{{ tag | handle }}" class="">{{ tag }}</a>
```
🔥 `class=""`

tauthomas01 replied:
I removed.

t-kelly commented on src/assets/styles/modules/blog-post.scss:
```diff
@@ -0,0 +1,15 @@
+  margin-bottom: 7px;
+}
+
+.blog-post__excerpt {
```
Don't think we need this style or class. This style is subjective and probably will change when you style your theme.

tauthomas01 replied:
I removed it.

t-kelly commented on src/assets/styles/modules/blog-post.scss:
```diff
@@ -0,0 +1,15 @@
+  grid-template-columns: 1fr 1fr 1fr;
+}
+
+.blog-post__date {
```
I think we can remove this style and class completely.

Can we use a `<p>` element for the date so by default, we get `display: block`? Also, I don't think we need the margin-bottom as it will probably change when styling the theme.

tauthomas01 replied:
I removed it and changed to `<p>` element.

t-kelly commented on src/sections/blog-post.liquid:
```diff
@@ -0,0 +1,155 @@
+  {% unless blog.articles_count == 0 %}
+    <div class="blog-post">
+      {% for article in blog.articles limit: section.settings.post_limit %}
+        <div class="blog-post__item">
```
We don't have any styles for this class, so may as well omit it?

tauthomas01 replied:
Good point.

t-kelly commented on src/sections/blog-post.liquid:
```diff
@@ -0,0 +1,155 @@
+            {% endif %}
+
+            {% if section.settings.blog_show_date %}
+              <span class="blog-post__date">
```
Remove this class since it's not needed with changes to CSS mentioned above.

t-kelly commented on src/sections/blog-post.liquid:
```diff
@@ -0,0 +1,155 @@
+              </span>
+            {% endif %}
+
+            <div class="blog-post__excerpt">
```
Remove this class since it's not needed with changes to CSS mentioned above.

tauthomas01 replied:
I removed.

t-kelly commented on src/sections/blog-post.liquid:
```diff
@@ -0,0 +1,155 @@
+      {% endfor %}
+    </div>
+  {% else %}
+    <div class="">
```
Is this `<div>` needed?

tauthomas01 replied:
I removed.

t-kelly commented on src/sections/blog-post.liquid:
```diff
@@ -0,0 +1,155 @@
+  {%- assign blog = blogs[section.settings.blog] -%}
+
+  {% unless blog.articles_count == 0 %}
+    <div class="blog-post">
```
Is this `<div>` needed? Maybe we can move this class to the `<section>` element?

tauthomas01 replied:
I moved it into <section> element and removed the `<div>`. I need to test the `else` statement and see if it doesn't break the layout.

t-kelly commented on src/sections/blog-post.liquid:
```diff
@@ -0,0 +1,155 @@
+  {% else %}
+    <div class="">
+      {% for i in (1..section.settings.post_limit) %}
+        <div class="">
```
🔥 `class=""` here and below.

tauthomas01 replied:
I removed.

t-kelly commented on src/assets/styles/modules/blog-post.scss:
```diff
@@ -0,0 +1,15 @@
+/* ================ Blog post ================ */
+.blog-post {
+  display: grid;
```
I feel like this approach might break if we increase the post count above 3. Maybe flex-box would be more accommodating?

tauthomas01 replied:
Yes, I use flexbox now. Since I don't have much experience with CSS grids, I wanted to give it a try and see how it works.

tauthomas01 commented:
@t-kelly I pushed my changes. I removed the styles and the `margin-bottom` on the date and excerpt. At first I feel like it was missing some minimal visual indicators so that's why I added some spacing.

tauthomas01 commented:
@t-kelly If you have time, can you check this PR? Thank you!

chrisberthe commented on src/assets/styles/sections/blog-post.scss:
```diff
@@ -0,0 +1,14 @@
+/* ================ Blog post ================ */
+.blog-post {
+  display: flex;
+  flex-direction: row;
```
Not needed, this is the default value.

chrisberthe commented on src/assets/styles/theme.scss:
```diff
@@ -34,3 +34,4 @@
 /* Sections */
 @import './sections/image-with-text';
 @import './sections/featured-collection';
+@import './sections/blog-post';
```
Nitpick: would you mind placing these in alphabetical order? Personal opinion, I find it easier to compare the list of files in the sidebar with the one in `theme.scss`.

chrisberthe commented on src/sections/blog-post.liquid:
```diff
@@ -0,0 +1,147 @@
+
+  Theme Store required settings
+  - Blog (which blog category to show in your store)
+  - Number of posts to show
```
The number of posts is an optional setting.

chrisberthe replied:
The heading, not listed here (but shown below as `section.settings.title`), is a required setting.

chrisberthe commented on src/sections/blog-post.liquid:
```diff
@@ -0,0 +1,147 @@
+{%- comment -%}
+  This is a required section for the Shopify Theme Store.
+  It is available when you add "Blog Posts" section on the Theme Editor.
```
>**in** the Theme Editor.

chrisberthe commented on src/sections/blog-post.liquid:
```diff
@@ -0,0 +1,147 @@
+
+  {%- comment -%}
+    This refer from blog section when you add the section from Theme Editor
+  {%- endcomment -%}
```
Do we need this comment? We have other instances where we're doing something similar for a selected `collection`, and not including a comment.

chrisberthe commented on src/sections/blog-post.liquid:
```diff
@@ -0,0 +1,147 @@
+{% schema %}
+  {
+    "name": "Blog posts",
+    "class": "index-section",
```
We can remove this class.

chrisberthe commented on src/sections/blog-post.liquid:
```diff
@@ -0,0 +1,147 @@
+  {%- endcomment -%}
+  {%- assign blog = blogs[section.settings.blog] -%}
+
+  {% unless blog.articles_count == 0 %}
```
We don't need this check - the for loop is good enough since it won't iterate if there aren't any blog articles. You'll need to add the `{% else %}` inside of this for loop though, e.g.:

```liquid
{% for article in blog.articles %}
  // output blog articles
{% else %}
  {% for i in (1..section.settings.post_limit) %}
      // output placeholders
  {% endfor %}
{% endfor %}

chrisberthe commented on src/sections/blog-post.liquid:
```diff
@@ -0,0 +1,147 @@
+{%- comment -%}
```
Can we rename this file to reflect the section name, i.e. `blog-posts.liquid`?

chrisberthe commented on src/sections/blog-post.liquid:
```diff
@@ -0,0 +1,147 @@
+    {% for i in (1..section.settings.post_limit) %}
+      <div>
+        <h3>
+          <a href="#">{{ 'homepage.onboarding.blog_title' | t }}</a>
```
We're missing translations for this onboarding content.

chrisberthe commented on src/sections/blog-post.liquid:
```diff
@@ -0,0 +1,147 @@
+        </h3>
+
+        {% if section.settings.blog_show_author %}
+          <span>{{ 'homepage.onboarding.blog_author' | t }}</span>
```
Same here.

chrisberthe commented on src/sections/blog-post.liquid:
```diff
@@ -0,0 +1,147 @@
+        {% endif %}
+
+        <div>
+          {{ 'homepage.onboarding.blog_excerpt' | t }}
```
Same here.

chrisberthe commented:
Getting linting errors for styles as well:

![image](https://user-images.githubusercontent.com/991693/39433963-3e226d32-4c65-11e8-8099-e57959942b77.png)

tauthomas01 commented:
@chrisberthe Thanks, I'm creating another PR to fix those issues.
